crates/bitcoin-da/src/helpers/parsers.rs

```diff
@@ -71,111 +84,109 @@ fn get_script(tx: &Transaction) -> Result<&Script, ParserError> {
+        return Err(ParserError::InvalidHeaderLength);
+    };
+
+    // Check rollup name
+    if header.rollup_name != rollup_name.as_bytes() {
+        return Err(ParserError::InvalidRollupName);
+    }
+
+    // Parse transaction body according to type
+    match header.typ {
+        TransactionType::Inscribed => parse_type_0_body(instructions),
+        TransactionType::Unknown(_) => Err(ParserError::InvalidHeaderType),
+    }
+}
+
+fn read_instr<'a>(
+    instructions: &mut dyn Iterator<Item = Result<Instruction<'a>, ParserError>>,
+) -> Result<Instruction<'a>, ParserError> {
+    let instr = instructions
+        .next()
+        .unwrap_or(Err(ParserError::UnexpectedEndOfScript))?;
+    Ok(instr)
+}
+
+fn read_push_bytes<'a>(
+    instructions: &mut dyn Iterator<Item = Result<Instruction<'a>, ParserError>>,
+) -> Result<&'a StructPushBytes, ParserError> {
+    let instr = read_instr(instructions)?;
+    match instr {
+        PushBytes(push_bytes) => Ok(push_bytes),
+        _ => Err(ParserError::UnexpectedOpcode),
+    }
+}
+
+fn read_opcode(
+    instructions: &mut dyn Iterator<Item = Result<Instruction<'_>, ParserError>>,
+) -> Result<Opcode, ParserError> {
+    let instr = read_instr(instructions)?;
+    let Op(op) = instr else {
+        return Err(ParserError::UnexpectedOpcode);
+    };
+    Ok(op)
+}
+
+// Parse transaction body of Type0
+fn parse_type_0_body(
+    instructions: &mut dyn Iterator<Item = Result<Instruction<'_>, ParserError>>,
+) -> Result<ParsedInscription, ParserError> {
+    // PushBytes(XOnlyPublicKey)
+    let _public_key = read_push_bytes(instructions)?;
+    if OP_CHECKSIG != read_opcode(instructions)? {
+        return Err(ParserError::UnexpectedOpcode);
+    }
+
+    let op_false = read_push_bytes(instructions)?;
+    if !op_false.is_empty() {
+        return Err(ParserError::UnexpectedOpcode);
+    }
+
+    if OP_IF != read_opcode(instructions)? {
+        return Err(ParserError::UnexpectedOpcode);
+    }
+
+    let signature = read_push_bytes(instructions)?;
+    let public_key = read_push_bytes(instructions)?;
+
+    let mut chunks = vec![];
+
+    loop {
+        let instr = read_instr(instructions)?;
+        match instr {
+            PushBytes(chunk) => chunks.push(chunk),
+            Op(OP_ENDIF) => break,
+            Op(_) => return Err(ParserError::UnexpectedOpcode),
         }
     }
 
-    if body.is_empty() || signature.is_empty() || public_key.is_empty() {
-        return Err(ParserError::EnvelopeHasIncorrectFormat);
+    // Nonce
+    let _nonce = read_push_bytes(instructions)?;
+    if OP_DROP != read_opcode(instructions)? {
+        return Err(ParserError::UnexpectedOpcode);
+    }
+    // END of transaction
+    if instructions.next().is_some() {
+        return Err(ParserError::UnexpectedOpcode);
+    }
+
+    let body_size: usize = chunks.iter().map(|c| c.len()).sum();
+    let mut body = Vec::with_capacity(body_size);
```

Comment:
are we really gonna benefit from creating a vec with capacity while we are iterating over chunks and summing lengths?

calculating length vs increasing the capacity of a vector.

wdyt? which has less cycles

Reply:
increasing the capacity of a vector on the fly requires reallocation which is not free. Storing refs of chunks, summing their length is almost a no op for risc0